pmd-core/src/main/java/net/sourceforge/pmd/cpd/AntlrTokenizer.java

```diff
@@ -0,0 +1,97 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.cpd;
+
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.Lexer;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+
+import net.sourceforge.pmd.lang.ast.TokenMgrError;
+
+/**
+ * Generic implementation of a {@link Tokenizer} useful to any Antlr grammar.
+ */
+public abstract class AntlrTokenizer implements Tokenizer {
```

Comment:
We are defining [an abstract class with no abstract methods](https://pmd.github.io/pmd-6.6.0/pmd_rules_java_bestpractices.html#abstractclasswithoutabstractmethod).

The class now also uses a single Lexer for all tokenizations (something that the Swift implementation didn't do, although I grant there is no hard requirement for that at the moment).

However, to do so, we must create `AntlrTokenizer` instances with `Lexer`s refering to `null` sources, which seems odd...

I think all 3 problemas have a very simple solution:
1. Have this class define a `protected abstract Lexer getLexerForSource(String source)`
2. Remote the constructor + `Lexer` field
3. Have subclasses implement the abstract method and use it locally from `tokenize`

Reply:
Given that the `Lexer` implementation requires a `CharStream` input we decided to define the method as: `protected abstract Lexer getLexerForSource(CharStream charStream);` to avoid duplicated code in every `AntlrTokenizer` subclass.